sql: update generated column content (#1109)
pingcap/docs: sql/generated-columns.md

TiDB supports generated columns as part of MySQL 5.7 compatibility. One of the primary use cases for generated columns is to extract data out of a JSON data type and enable it to be indexed.

> ## Index JSON using generated stored column

Reviewer: ```suggestion
## Index JSON using stored generated column
```

);
```

To index a JSON column, you must first extract it as a generated stored column. 

> **Note:**
>

> > The optimizer in TiDB only uses the index created on a generated stored column. Currently, the optimizer cannot use an index created on a generated virtual column. This issue will be fixed in the later TiDB versions (See [#5189](https://github.com/pingcap/tidb/issues/5189)).

Reviewer 2: ```suggestion
> The optimizer in TiDB only uses the index created on a generated stored column. Currently, the optimizer cannot use an index created on a generated virtual column. This issue will be fixed in later TiDB versions (See [#5189](https://github.com/pingcap/tidb/issues/5189)).
```

> > The optimizer in TiDB only uses the index created on a generated stored column. Currently, the optimizer cannot use an index created on a generated virtual column. This issue will be fixed in the later TiDB versions (See ISSUE [#5189](https://github.com/pingcap/tidb/issues/5189)).

Reviewer: ```suggestion
> The optimizer in TiDB only uses the index created on a generated stored column. Currently, the optimizer cannot use an index created on a generated virtual column. This issue will be fixed in the later TiDB versions (See [#5189](https://github.com/pingcap/tidb/issues/5189)).
```

> Using the `city` generated stored column as an example, you are then able to add an index:

Reviewer: Sorry, one more: 
```suggestion
Using the `city` stored generated column as an example, you are then able to add an index:
```

```sql
CREATE TABLE person (
    id INT NOT NULL AUTO_INCREMENT PRIMARY KEY,
    name VARCHAR(255) NOT NULL,
    address_info JSON,
    city VARCHAR(64) AS (JSON_UNQUOTE(JSON_EXTRACT(address_info, '$.city'))) STORED,
    KEY (city)
);
```

> In this table, the `city` column is a **generated column**. As the name implies, the column is generated from other columns in the table, and cannot be assigned a value when inserted or updated. This column is generated based on the expression it defines and it is stored in the database. Thus this column can be read directly, not in a way that the `address_info` column it depends on is read first and then the data is calculated. The index on `city` however is _stored_ and uses the same structure as other indexes of the type `varchar(64)`.

Reviewer 2: ```suggestion
In this table, the `city` column is a **generated column**. As the name implies, the column is generated from other columns in the table, and cannot be assigned a value when inserted or updated. This column is generated based on a defined expression and is stored in the database. Thus this column can be read directly, not in a way that its dependent column `address_info`  is read first and then the data is calculated. The index on `city` however is _stored_ and uses the same structure as other indexes of the type `varchar(64)`.
```

> You can use the index on the generated stored column in order to speed up the following statement:

Reviewer: ```suggestion
You can use the index on the stored generated column in order to speed up the following statement:
```

ERROR 1048 (23000): Column 'city' cannot be null
```

## Use generated virtual columns

> TiDB also supports generated virtual columns. Different from generated store columns, generated virtual columns are **virtual** in that they are generated as needed and are not stored in the database or occupied the memory space. Although TiDB supports indexing generated virtual columns, the optimizer currently cannot use the index in this case. This issue will be fixed in a later version of TiDB ([#5189](https://github.com/pingcap/tidb/issues/5189)).

Reviewer 2: ```suggestion
TiDB also supports generated virtual columns. Different from generated store columns, generated virtual columns are **virtual** in that they are generated as needed and are not stored in the database or cached in the memory. Although TiDB supports indexing generated virtual columns, the optimizer currently cannot use indexes in this case. This issue will be fixed in a later version of TiDB ([#5189](https://github.com/pingcap/tidb/issues/5189)).
```

> TiDB also supports generated virtual columns. Different from generated store columns, generated virtual columns are **virtual** in that they are generated as needed and are not stored in the database or occupied the memory space. Although TiDB supports indexing generated virtual columns, the optimizer currently cannot use the index in this case. This issue will be fixed in the later TiDB versions. (ISSUE [#5189](https://github.com/pingcap/tidb/issues/5189)).

Reviewer: ```suggestion
TiDB also supports generated virtual columns. Different from generated store columns, generated virtual columns are **virtual** in that they are generated as needed and are not stored in the database or occupied the memory space. Although TiDB supports indexing generated virtual columns, the optimizer currently cannot use the index in this case. This issue will be fixed in a later version of TiDB ([#5189](https://github.com/pingcap/tidb/issues/5189)).
```
